newrelic/resource_newrelic_service_level.go

```diff
@@ -0,0 +1,302 @@
+package newrelic
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/newrelic/newrelic-client-go/pkg/common"
+	"github.com/newrelic/newrelic-client-go/pkg/errors"
+)
+
+func resourceNewRelicServiceLevel() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceNewRelicServiceLevelCreate,
+		ReadContext:   resourceNewRelicServiceLevelRead,
+		UpdateContext: resourceNewRelicServiceLevelUpdate,
+		DeleteContext: resourceNewRelicServiceLevelDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"guid": {
+				Type:         schema.TypeString,
+				Required:     true,
+				Description:  "",
+				ValidateFunc: validation.StringIsNotWhiteSpace,
+			},
+			"name": {
+				Type:         schema.TypeString,
+				Required:     true,
+				Description:  "",
+				ValidateFunc: validation.StringIsNotWhiteSpace,
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "",
+			},
+			"events": {
+				Type:        schema.TypeList,
+				Required:    true,
+				Description: "",
+				MinItems:    1,
+				MaxItems:    1,
+				Elem:        eventsSchema(),
+			},
+			"objective": {
+				Type:        schema.TypeSet,
```

Comment:
Not a blocker by any means, but out of curiosity, could this also be `schema.TypeList`? I lack the domain knowledge of the `serviceLevel` feature, and the main reason I ask is because `TypeSet` has given us some "interesting" behavior to deal with in some scenarios, where often times a `TypeList` works just fine and is a little easier to work with when evaluating state changes.

Reply:
Also the field in the GraphQL mutation is plural `objectives`. We've typically tried to make sure Terraform attributes match the API fields to continuity across products 🙂

Reply:
Hello @sanderblue, thanks for the feedback!

Regarding the `TypeSet`, ideally I'd prefer to model it as a set because it wouldn't make much sense to have two objectives that are exactly the same. For instance, you would want one objective that is "realistic" with a target of 95% and another one that is "optimistic" with a target of 99%.

Regarding the naming of the `objectives`, as far as I could see in the documentation, we usually use a singular name for blocks, for instance, with [Entity tags](https://registry.terraform.io/providers/newrelic/newrelic/latest/docs/resources/entity_tags#tag) or [Workload search queries](https://registry.terraform.io/providers/newrelic/newrelic/latest/docs/resources/workload#entity_search_query). In fact, the naming in singular sounds more natural to me considering that you can define multiple blocks, eg:

```
resource "newrelic_service_level" "foo" {
   ...
   objective { ... }
   objective { ... }
}
```

Nevertheless, I am happy to change it if you think that pluralizing the name is the recommended approach 🙂

Reply:
Fair enough. Thanks for the thought process behind the decision. Looks good to me.